pkg/block/fetcher_test.go

```diff
@@ -393,3 +401,271 @@ func TestTimePartitionMetaFilter_Filter(t *testing.T) {
+							Sources: []ulid.ULID{ULID(3)},
+						},
+					},
+				},
+				ULID(4): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(4),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
+						},
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
+				return map[ulid.ULID]*metadata.Meta{
+					ULID(4): input[ULID(4)],
+				}
+			},
+		},
+		{
+			name: "3 compacted metas of level 2 and compacted meta of level 3 in bucket",
+			input: map[ulid.ULID]*metadata.Meta{
+				ULID(10): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(10),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3)},
+						},
+					},
+				},
+				ULID(11): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(11),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(4), ULID(5), ULID(6)},
+						},
+					},
+				},
+				ULID(12): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(12),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(7), ULID(8), ULID(9)},
+						},
+					},
+				},
+				ULID(13): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(13),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4), ULID(5), ULID(6), ULID(7), ULID(8), ULID(9)},
+						},
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
+				return map[ulid.ULID]*metadata.Meta{
+					ULID(13): input[ULID(13)],
+				}
+			},
+		},
+		{
+			name: "compacted blocks with overlapping sources",
+			input: map[ulid.ULID]*metadata.Meta{
+				ULID(1): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(1),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1)},
+						},
+					},
+				},
+				ULID(5): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(5),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2)},
+						},
+					},
+				},
+				ULID(6): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(6),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
+						},
+					},
+				},
+				ULID(7): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(7),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3)},
+						},
+					},
+				},
+				ULID(8): {
+					BlockMeta: tsdb.BlockMeta{
+						ULID: ULID(8),
+						Compaction: tsdb.BlockMetaCompaction{
+							Sources: []ulid.ULID{ULID(1), ULID(2), ULID(3), ULID(4)},
+						},
+					},
+				},
+			},
+			expected: func(input map[ulid.ULID]*metadata.Meta) map[ulid.ULID]*metadata.Meta {
+				return map[ulid.ULID]*metadata.Meta{
+					ULID(6): input[ULID(6)],
+				}
+			},
+		},
+	} {
+		f := NewCompactedBlockFilter()
```

Comment:
Can we add case with 2 blocks having exactly the same sources?

Reply:
Yes, its added with test case name: `two compacted blocks with same sources`